### PR TITLE
fix(expr): don't panic on table function errors

### PR DESCRIPTION
**Pull request by wangrunji0408**
Signed-off-by: Runji Wang <wangrunji0408@163.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

partially fix_ #11915 

Currently, when an error occurs in a table function call, the returned stream stops and the streaming actor will exit. This PR makes streaming tolerant to table function errors. When an error occurs, an error column will be appended to the returned chunk. (see the example in [arrow-udf](https://github.com/risingwavelabs/arrow-udf/tree/main/arrow-udf#fallible-functions)) The batch side will raise error immediately when seeing this column, while the streaming side will report errors and keep going. If some rows have been output when the error occurs, these rows will not be reverted.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [x] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/expr/macro/src/gen.rs [hunk 1]
```diff
@@ -1122,11 +1122,29 @@ impl FunctionAttr {
...
+            ReturnTypeKind::Option => quote! { match #iter {
+                Some(it) => it,
+                None => continue,
+            } },
+            ReturnTypeKind::Result => quote! { match #iter {
+                Ok(it) => it,
+                Err(e) => {
+                    index_builder.append(Some(i as i32));
+                    #(#builders.append_null();)*
+                    error_builder.append(Some(&e.to_string()));
```
Comment (BugenZhao) [inline src/expr/macro/src/gen.rs:1134]: Shall we use `as_report` and `Utf8ArrayBuilder::writer`?

File: src/expr/macro/src/gen.rs [hunk 2]
```diff
@@ -1223,17 +1255,19 @@ impl FunctionAttr {
...
-                                }
+                                #append_output
 
                                 if index_builder.len() == self.chunk_size {
                                     let len = index_builder.len();
                                     let index_array = std::mem::replace(&mut index_builder, I32ArrayBuilder::new(self.chunk_size)).finish().into_ref();
                                     let value_arrays = [#(std::mem::replace(&mut #builders, #builder_types::with_type(self.chunk_size, #return_types)).finish().into_ref()),*];
                                     #build_value_array
-                                    yield DataChunk::new(vec![index_array, value_array], self.chunk_size);
+                                    let error_array = std::mem::replace(&mut error_builder, Utf8ArrayBuilder::new(self.chunk_size)).finish().into_ref();
```
Comment (BugenZhao) [inline src/expr/macro/src/gen.rs:1265]: I think there might be no need to reserve capacity for the error message array builder. Also the implementation of reservation for `Utf8ArrayBuilder` makes little sense. 🤣

https://github.com/risingwavelabs/risingwave/blob/d19d8f9ae55832430f452c2ed90fca44910dd37b/src/common/src/array/bytes_array.rs#L142-L146

File: e2e_test/streaming/bug_fixes/issue_11915.slt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+create materialized view mv as select x, generate_series(1, 2, x) from t;
+
+# x = 0 causes generate_series(1, 2, x) to return an error.
+statement ok
+insert into t values (0), (1);
+
+statement ok
+flush;
+
+# Output 0 row when the set-returning function returns error.
```
Comment (BugenZhao) [inline e2e_test/streaming/bug_fixes/issue_11915.slt:16]: There was an discussion on the behavior in this case. https://github.com/risingwavelabs/risingwave/issues/12474#issuecomment-1763368554 cc @fuyufjh 

Not sure which one we should adopt.

Comment (wangrunji0408) [reply e2e_test/streaming/bug_fixes/issue_11915.slt:16]: The default behavior for a table function not being called (e.g. with null input) is returning no rows. So I vote for this for the sake of consistency.

File: src/expr/macro/src/gen.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/expr/macro/src/gen.rs:1265]: Actually we should distinguish `item_capacity` and `data_capacity` like [arrow](https://docs.rs/arrow/latest/arrow/array/builder/struct.GenericByteBuilder.html#tymethod.with_capacity) does. If only one capacity can be specified, it should refer to `item_capacity`. And here we expect `item_capacity = chunk_size` and `data_capacity = 0`.

Comment (wangrunji0408) [reply src/expr/macro/src/gen.rs:1265]: fixed in 2c52ee3

Comment (BugenZhao) [reply src/expr/macro/src/gen.rs:1265]: You're right. I just realized that even there's no error in the result, we still need to allocate the `offset` vector.

**Review by BugenZhao — APPROVED** [review]
LGTM!

File: src/expr/macro/src/gen.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/expr/macro/src/gen.rs:1265]: Ideally, if there is no error, we can avoid building the error array at all. But that would make the code much more complicated.

**Review by xxchan — COMMENTED** [review]
Thanks. Cool idea 👍

What I don't fully understand is where is the line leading to panic before? 🤣

> When an error occurs, an error column will be appended to the returned chunk. (see the example in [arrow-udf](https://github.com/risingwavelabs/arrow-udf/tree/main/arrow-udf?rgh-link-date=2024-06-06T09%3A55%3A43Z#fallible-functions))

To clarify, `arrow_udf::function` and `risingwave_expr::function` don't have strict relationship, but share similar ideas. Do I understand it correctly?

File: e2e_test/streaming/bug_fixes/issue_11915.slt (hunk 3 above)
Comment (fuyufjh) [reply e2e_test/streaming/bug_fixes/issue_11915.slt:16]: > The default behavior for a table function not being called (e.g. with null input) is returning no rows. So I vote for this for the sake of consistency.

I agree that these 2 cases should be consistent

**Comment by wangrunji0408** [comment]
> > When an error occurs, an error column will be appended to the returned chunk. (see the example in [arrow-udf](https://github.com/risingwavelabs/arrow-udf/tree/main/arrow-udf?rgh-link-date=2024-06-10T02%3A37%3A35Z#fallible-functions))
> 
> To clarify, `arrow_udf::function` and `risingwave_expr::function` don't have strict relationship, but share similar ideas. Do I understand it correctly?

Yes. They share the same behavior behind the interface so that we don't need to write additional adapters for UDFs.

File: e2e_test/streaming/bug_fixes/issue_11915.slt (hunk 3 above)
Comment (xiangjinwu) [reply e2e_test/streaming/bug_fixes/issue_11915.slt:16]: This is Option 1 in the original issue. And I agree it is consistent with null input. In other words, the table function returns `null` (which is similar to `{}` but not `{null}`) on error.

Related edges cases on the differences among `null`, `{}` and `{null}`:
* `array_agg` function on zero input rows return no rows with `group by`, but a `null` row with scalar agg. #15590
* Scalar subquery returning zero rows result in a `null` row. Array constructing subquery returning zero rows result in a `{}`. #15593

^ Just mentioning these cases. Not suggesting any value is always preferred than the other 2.

**Pull request merged**